Stream's structuredCloneForStream should keep resizability of ArrayBuffers

Constellation opened the pull request:
#### afcc2ab7b2c2fbdf5b97a179e204399a69d43bcc
<pre>
Stream&apos;s structuredCloneForStream should keep resizability of ArrayBuffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=248368">https://bugs.webkit.org/show_bug.cgi?id=248368</a>
rdar://102685044

Reviewed by Ross Kirsling.

Apply the structuredClone&apos;s change to structuredCloneForStream too.
This should keep ArrayBuffer / TypedArray&apos;s resizability.

* LayoutTests/http/wpt/fetch/request-clone-resizable-expected.txt: Added.
* LayoutTests/http/wpt/fetch/request-clone-resizable.html: Added.
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/StructuredClone.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/257043@main">https://commits.webkit.org/257043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08d51587c11238155e44b303ac08eddecdbeec03

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6924 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107174 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167440 "Found 1 new test failure: fast/images/pagewide-play-pause-offscreen-animations.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7313 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35690 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90072 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103829 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5485 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84310 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32472 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87365 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89131 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75392 "Found 3 new API test failures: /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/js-exception, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/network-error (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88584 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/919 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84255 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/910 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22061 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28516 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5728 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44522 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87085 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2397 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2161 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41453 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19538 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/WebCore/bindings/js/StructuredClone.cpp:
```diff
@@ -77,29 +77,65 @@ JSC_DEFINE_HOST_FUNCTION(structuredCloneForStream, (JSGlobalObject* globalObject
     JSValue value = callFrame->uncheckedArgument(0);
 
-    if (value.inherits<JSArrayBuffer>())
-        RELEASE_AND_RETURN(scope, cloneArrayBufferImpl(globalObject, callFrame, CloneMode::Full));
```
I take it that `cloneArrayBufferImpl` needs to stay as it is?

Constellation replied:
Right now, we leave `cloneArrayBufferImpl` since it is only used by ReadableByteStream, which is incomplete work, so we don't know whether this change is necessary for that.

rkirsling commented on Source/WebCore/bindings/js/StructuredClone.cpp:
```diff
@@ -77,29 +77,65 @@ JSC_DEFINE_HOST_FUNCTION(structuredCloneForStream, (JSGlobalObject* globalObject
 #define CLONE_TYPED_ARRAY(name) \
         case Type##name: { \
-            RELEASE_AND_RETURN(scope, JSValue::encode(JS##name##Array::create(globalObject, globalObject->typedArrayStructure(Type##name, isResizableOrGrowableShared), WTFMove(bufferClone), bufferView->byteOffset(), bufferView->length()))); \
+            RELEASE_AND_RETURN(scope, JSValue::encode(toJS(globalObject, globalObject, name##Array::wrappedAs(bufferClone.releaseNonNull(), byteOffset, length).get()))); \
```
Passing `globalObject` twice is interesting; looks like `SimpleTypedArrayController::toJS` even has `UNUSED_PARAM` for the first one, but is this necessary for API consistency somehow?

Constellation replied:
I think we can improve it in a future patch :)

rkirsling reviewed (approved):
r=me with questions

Constellation committed:
Stream's structuredCloneForStream should keep resizability of ArrayBuffers
https://bugs.webkit.org/show_bug.cgi?id=248368
rdar://102685044

Reviewed by Ross Kirsling.

Apply the structuredClone's change to structuredCloneForStream too.
This should keep ArrayBuffer / TypedArray's resizability.

* LayoutTests/http/wpt/fetch/request-clone-resizable-expected.txt: Added.
* LayoutTests/http/wpt/fetch/request-clone-resizable.html: Added.
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
* Source/WebCore/bindings/js/StructuredClone.cpp:
(WebCore::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/257043@main
